mlir/lib/Dialect/Rock/Transforms/VectorizeFusions.cpp

```diff
@@ -0,0 +1,132 @@
+#include "mlir/Dialect/Rock/Passes.h.inc"
+} // namespace rock
+} // namespace mlir
+
+#define DEBUG_TYPE "rock-vectorize-fusions"
+
+using namespace mlir;
+using mlir::gpu::AddressSpace;
+
+template <typename MemrefTypedValue>
+static AddressSpace getAddressSpace(MemrefTypedValue val) {
+  if (val.getType().getMemorySpace()) {
+    return val.getType()
+        .getMemorySpace()
+        .template cast<gpu::AddressSpaceAttr>()
+        .getValue();
+  }
+  return gpu::AddressSpace::Global;
+}
+
+namespace {
+struct RockVectorizeFusionsPass
+    : public rock::impl::RockVectorizeFusionsPassBase<
+          RockVectorizeFusionsPass> {
+  void runOnOperation() override;
+};
+} // end namespace
+
+void RockVectorizeFusionsPass::runOnOperation() {
+  func::FuncOp op = getOperation();
+  IRRewriter b(op.getContext());
+  op.walk([&](affine::AffineForOp loop) -> WalkResult {
+    // Collect data types
+    SmallVector<Type> loopTypes;
+    WalkResult canVectorize =
+        loop.getLoopBody().walk([&loopTypes](Operation *op) -> WalkResult {
+          if (auto affineLoad = dyn_cast<affine::AffineLoadOp>(op)) {
+            if (getAddressSpace(affineLoad.getMemref()) ==
+                AddressSpace::Private)
+              loopTypes.push_back(affineLoad.getType());
+          } else if (auto affineStore = dyn_cast<affine::AffineStoreOp>(op)) {
+            if (getAddressSpace(affineStore.getMemref()) ==
+                AddressSpace::Private)
+              loopTypes.push_back(affineStore.getMemRefType().getElementType());
+          } else if (dyn_cast<affine::AffineForOp>(op)) {
+            return WalkResult::interrupt();
+          }
+          return WalkResult::advance();
+        });
+
+    if (canVectorize.wasInterrupted())
+      return WalkResult::advance();
+
+    if (loopTypes.empty())
+      return WalkResult::advance();
+
+    LLVM_DEBUG(llvm::dbgs() << "Try to vectorize: " << loop << "\n");
+    const int64_t ub = loop.getConstantUpperBound();
+    const int64_t lb = loop.getConstantLowerBound();
+    const int64_t step = loop.getStep();
+    if (step > 1)
+      return WalkResult::advance();
+    const int64_t loopTripCount = (ub - lb);
+    const int64_t maxVectorBitWidth = 64;
+
+    // Look for the vectorization factor
+    int64_t vectorizationFactor = loopTripCount;
+    for (auto type : loopTypes)
+      vectorizationFactor =
+          math_util::gcd(maxVectorBitWidth / type.getIntOrFloatBitWidth(),
+                         vectorizationFactor);
+
+    if (vectorizationFactor == 1)
+      return WalkResult::advance();
+
+    LLVM_DEBUG(llvm::dbgs()
+               << "With vector length = " << vectorizationFactor << "\n");
+    DenseSet<Operation *> loops{loop};
+    affine::vectorizeAffineLoops(op, loops,
+                                 SmallVector<int64_t>{vectorizationFactor}, {});
+
+    // Make sure the transfer reads are in bounds
+    return WalkResult::advance();
+  });
+  op.walk([&](vector::TransferReadOp op) {
```

Comment:
There's a widget something like `getFilteredOps` that saves us the recursion of `walk()` which we don't need

Reply:
I used `getOps<optType>`, but the vectorization is destructive, hence I still have to walk through the affineForOps (and then filter out the transfer{read,writes})